{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "94f37cbe-8493-4800-9d01-71d263d116a8",
   "metadata": {},
   "source": [
    "This analysis was performed succesfully in \"CentOS Linux release 7.5.1804 (Core)\". If you use different version of Linux or OS, you may need to adjust it accordingly and we are not responsible for any liability or error.\n",
    "\n",
    "Other information and issue reporting, please refer to https://github.com/muharif/LiverAging_2022"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a9db06e-a9c1-4539-bade-eaea56a96404",
   "metadata": {},
   "source": [
    "## Init"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb1cf55f-272f-4b18-8136-a81f34a1f577",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "import rpy2,os,re, shutil\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import copy\n",
    "sns.set(font=\"Arial\")\n",
    "sns.set_style(\"white\")\n",
    "import matplotlib\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "from sklearn.decomposition import PCA as sklearnPCA\n",
    "os.environ['KMP_DUPLICATE_LIB_OK']='True'\n",
    "import pickle\n",
    "from statannot import add_stat_annotation\n",
    "\n",
    "#rpy2\n",
    "import rpy2.robjects as robjects\n",
    "import rpy2.robjects.numpy2ri\n",
    "rpy2.robjects.numpy2ri.activate()\n",
    "from rpy2.robjects import pandas2ri\n",
    "pandas2ri.activate()\n",
    "import warnings\n",
    "from rpy2.rinterface import RRuntimeWarning\n",
    "warnings.filterwarnings(\"ignore\", category=RRuntimeWarning)\n",
    "\n",
    "from scipy.stats import zscore, spearmanr\n",
    "from scipy.interpolate import interp1d\n",
    "from statsmodels.sandbox.stats.multicomp import multipletests\n",
    "import leidenalg\n",
    "import igraph as ig\n",
    "import gseapy as gp\n",
    "from scipy.cluster.hierarchy import dendrogram, linkage, cut_tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "604def81-3c66-4f89-b2a5-e83220657297",
   "metadata": {},
   "outputs": [],
   "source": [
    "class DESeq_Piano:\n",
    "    def __init__(self,respath='.'):\n",
    "        if respath=='.':\n",
    "            print('Result Path is Not Defined, any result will be written in this directory')\n",
    "            self.respath=respath\n",
    "        elif os.path.isdir(respath):\n",
    "            print('Result Path Exists, please be careful of overwriting')\n",
    "            self.respath=respath\n",
    "        else:\n",
    "            print('Result Path Does Not Exists, creating new directory')\n",
    "            os.mkdir(respath)\n",
    "            self.respath=respath\n",
    "        if os.path.isdir(respath+'/Figures'):\n",
    "            pass\n",
    "        else:\n",
    "            os.mkdir(respath+'/Figures')\n",
    "        self.__Init()\n",
    "\n",
    "            \n",
    "    def __Init(self):\n",
    "\n",
    "        \n",
    "        robjects.r('''\n",
    "            library('DESeq2')\n",
    "            library('piano')\n",
    "            library('Biobase')\n",
    "            library('snow')\n",
    "            library('RColorBrewer')\n",
    "            library('gplots')\n",
    "            library('snowfall')\n",
    "        ''')\n",
    "        self.GSC_KEGG=None\n",
    "        self.GSC_GO=None\n",
    "        self.KEGG=None\n",
    "        self.GO=None\n",
    "        self.dds=None\n",
    "        self.deseq_res=None\n",
    "        self.conds=None\n",
    "        self.count=None\n",
    "        self.tpm=None\n",
    "\n",
    "    def __check_dir(self,extrapath):\n",
    "        if os.path.isdir('%s/%s/' % (self.respath,extrapath)):\n",
    "            print('Path %s/%s/ exist, writing in the folder' % (self.respath,extrapath))\n",
    "        else:\n",
    "            print('Path %s/%s/ does not exist, making the folder to write the results' % (self.respath,extrapath))\n",
    "            os.mkdir('%s/%s/' % (self.respath,extrapath))\n",
    "            \n",
    "    def DEseq(self,count_df=None,conds_series=None):\n",
    "        if count_df is not None:\n",
    "            self.count=count_df\n",
    "        if conds_series is not None:\n",
    "            self.conds=conds_series\n",
    "        self.gene_names=count_df.index\n",
    "        robjects.globalenv['conds']=np.array(self.conds)\n",
    "        robjects.globalenv['subject']=np.array(self.conds.index)\n",
    "        robjects.globalenv['count']=self.count.astype(int).values\n",
    "        robjects.r('''\n",
    "            conds=as.factor(conds)\n",
    "            coldata <- data.frame(row.names=subject,conds)\n",
    "            dds <- DESeqDataSetFromMatrix(countData=as.matrix(count),colData=coldata,design=~conds)\n",
    "            dds <- DESeq(dds)\n",
    "            print(resultsNames(dds))\n",
    "            #baseMeanPerLvl <- sapply( levels(conds), function(lvl) rowMeans( counts(dds,normalized=TRUE)[,conds == lvl] ) )\n",
    "            #conds2=colnames(baseMeanPerLvl)\n",
    "            #gene_names=rownames(baseMeanPerLvl)\n",
    "\n",
    "        ''')\n",
    "        self.dds=robjects.globalenv['dds']\n",
    "    \n",
    "    def DEseq_Compare(self, cond1, cond2,save=True):\n",
    "        if self.dds is None:\n",
    "            raise ValueError('No DESeq data found, run the DEseq function')\n",
    "        else:\n",
    "            print('Comparing %s and %s' % (cond1,cond2))\n",
    "            robjects.globalenv['dds']=self.dds\n",
    "            robjects.globalenv['cond1']=cond1\n",
    "            robjects.globalenv['cond2']=cond2\n",
    "            robjects.globalenv['gene_names']=np.array(self.gene_names)\n",
    "            robjects.r('''\n",
    "                res=results(dds,contrast=c('conds',cond1,cond2))\n",
    "                res=data.frame(res)\n",
    "                res$rownames=gene_names\n",
    "            ''')\n",
    "            res = pd.DataFrame(robjects.globalenv['res']).set_index('rownames')\n",
    "            self.deseq_df=res\n",
    "            if save:\n",
    "                self.__check_dir('deseq')\n",
    "                res.to_csv('%s/deseq/deseq_%s_%s.txt' % (self.respath,cond1,cond2),sep='\\t')\n",
    "        \n",
    "    def loadGSC(self,KEGG='',TF='',RM='',GO='',extraGSC=''):\n",
    "        if KEGG != '':\n",
    "            name='KEGG'\n",
    "            self.__check_dir(name)\n",
    "            self.GSC_KEGG=self.__loadGSC_execute(KEGG)\n",
    "        if GO != '':\n",
    "            name='GO'\n",
    "            self.__check_dir(name)\n",
    "            self.GSC_GO=self.__loadGSC_execute(GO)\n",
    "\n",
    "    \n",
    "    def __loadGSC_execute(self,GSC):\n",
    "        robjects.globalenv['GSC']=GSC \n",
    "        robjects.r('''\n",
    "            y=loadGSC(GSC)\n",
    "        ''')\n",
    "        return robjects.globalenv['y']\n",
    "    \n",
    "    def __PIANO_execute(self,cond1,cond2,GSCtype='',save=True):\n",
    "        robjects.globalenv['deseq_file']=pandas2ri.py2rpy_pandasdataframe(self.deseq_df)\n",
    "        if GSCtype == 'KEGG':\n",
    "            robjects.globalenv['y']=self.GSC_KEGG\n",
    "            deseq_df=self.deseq_df\n",
    "            robjects.globalenv['deseq_file']=pandas2ri.py2rpy_pandasdataframe(deseq_df)\n",
    "        elif GSCtype == 'GO':\n",
    "            robjects.globalenv['y']=self.GSC_GO\n",
    "            deseq_df=self.deseq_df\n",
    "            deseq_df.index=[i.upper() for i in deseq_df.index]\n",
    "            robjects.globalenv['deseq_file']=pandas2ri.py2rpy_pandasdataframe(deseq_df)\n",
    "        robjects.r('''\n",
    "            DESeq_file=deseq_file\n",
    "            DESeq_file=DESeq_file[ ,c('log2FoldChange','pvalue')]\n",
    "            logFC=as.matrix(DESeq_file[,1])\n",
    "            pval=as.matrix(DESeq_file[,2])\n",
    "            rownames(logFC)=(rownames(DESeq_file))\n",
    "            rownames(pval)=(rownames(DESeq_file))\n",
    "            logFC[is.na(logFC)] <- 0\n",
    "            pval[is.na(pval)] <- 1\n",
    "            gsaRes <- runGSA(pval,logFC,gsc=y, geneSetStat=\"reporter\", signifMethod=\"nullDist\", nPerm=1000,ncpus=8,gsSizeLim = c(5, Inf))\n",
    "            res_piano=GSAsummaryTable(gsaRes)\n",
    "            res_piano$rownames=rownames(res_piano)\n",
    "        ''')\n",
    "        res=pd.DataFrame(robjects.globalenv['res_piano']).set_index('rownames').set_index('Name').iloc[0:,0:]#.groupby('Name').sum()\n",
    "        if save:\n",
    "            res.to_csv('%s/%s/piano_%s_%s.txt' % (self.respath,GSCtype,cond1,cond2),sep='\\t')\n",
    "        return res\n",
    "\n",
    "    \n",
    "    def PIANO(self,cond1,cond2,heatmap=True,save=True):\n",
    "        if save:\n",
    "            self.__check_dir('Figures')\n",
    "        if self.deseq_df is None:\n",
    "            raise ValueError('No DEseq comparison found, run DEseq_Compare first or assign deseq result dataframe to deseq_res')\n",
    "        else:\n",
    "            if self.GSC_KEGG is not None:\n",
    "                name='KEGG'\n",
    "                self.KEGG = self.__PIANO_execute(cond1,cond2,GSCtype=name,save=save)\n",
    "            if self.GSC_GO is not None:\n",
    "                name='GO'\n",
    "                self.GO = self.__PIANO_execute(cond1,cond2,GSCtype=name,save=save)       \n",
    "                    \n",
    "    def __piano2xlsx(self,name,part='ALL'):\n",
    "        if part != 'ALL':\n",
    "            files=[i for i in sorted(os.listdir('%s/%s' % (self.respath,name))) if part in i]\n",
    "        else:\n",
    "            files=sorted(os.listdir('%s/%s' % (self.respath,name)))\n",
    "        writer = pd.ExcelWriter('%s/%s_%s.xlsx' % (self.respath,name,part), engine='xlsxwriter')\n",
    "        for i in files:\n",
    "            if i.startswith('.'):\n",
    "                continue\n",
    "            deseq1=pd.read_csv('%s/%s/%s' % (self.respath,name,i),index_col='Name',sep='\\t')[['Genes (tot)','Stat (dist.dir.up)','p adj (dist.dir.up)']]\n",
    "            deseq1.columns=['# of Genes', 'Stats','P-Adj']\n",
    "            deseq1=deseq1[deseq1['P-Adj']<0.05]\n",
    "            deseq1['Direction']='UP'\n",
    "            deseq2=pd.read_csv('%s/%s/%s' % (self.respath,name,i),index_col='Name',sep='\\t')[['Genes (tot)','Stat (dist.dir.up)','p adj (dist.dir.dn)']]\n",
    "            deseq2.columns=['# of Genes', 'Stats','P-Adj']\n",
    "            deseq2=deseq2[deseq2['P-Adj']<0.05]\n",
    "            deseq2['Direction']='DOWN'\n",
    "            deseq1=pd.concat([deseq1,deseq2])\n",
    "            if len(i.replace('piano_','').replace('.txt',''))>31:\n",
    "                print('sheet name too big, trimming to 30 character')\n",
    "                n_temp=i.replace('piano_','').replace('.txt','')[0:31]\n",
    "            else:\n",
    "                n_temp=i.replace('piano_','').replace('.txt','')\n",
    "            deseq1.sort_values('P-Adj').to_excel(writer, sheet_name=n_temp)\n",
    "        writer.save()\n",
    "        \n",
    "    def summarizePiano(self,names=['KEGG','GO','TF','RM','extraGSC'],part='ALL'):\n",
    "        for name in names:\n",
    "            if os.path.isdir(self.respath+name):\n",
    "                self.__piano2xlsx(name,part=part)\n",
    "            else:\n",
    "                continue\n",
    "    \n",
    "    def summarizeDEseq(self,deseq_result='deseq',part='ALL'):\n",
    "        if os.path.isdir(self.respath+deseq_result):\n",
    "            if part != 'ALL':\n",
    "                files=[i for i in sorted(os.listdir(self.respath+'/deseq/')) if part in i]\n",
    "            else:\n",
    "                files=sorted(os.listdir(self.respath+'/deseq/'))\n",
    "            writer = pd.ExcelWriter(self.respath+'/DifferentialExpression_%s.xlsx' % part, engine='xlsxwriter')\n",
    "            fin=0\n",
    "            for i in files:\n",
    "                if i.startswith('.'):\n",
    "                    continue\n",
    "                deseq=pd.read_csv(self.respath+'/deseq/%s' % i,index_col=0,sep='\\t')\n",
    "                deseq['abs']=deseq['log2FoldChange'].abs()\n",
    "                deseq=deseq.sort_values('abs',ascending=False)[['log2FoldChange','pvalue','padj']]\n",
    "\n",
    "                deseq['Direction']=['UP' if i > 0 else 'DOWN' for i in deseq['log2FoldChange']]\n",
    "                deseq.columns=['L2FC','P-VALUE','P-ADJ','DIRECTION']\n",
    "                deseq=deseq.sort_values('P-VALUE')\n",
    "                if len(i.replace('deseq_','').replace('.txt',''))>31:\n",
    "                    print('sheet name too big, trimming to 30 character')\n",
    "                    n_temp=i.replace('deseq_','').replace('.txt','')[0:31]\n",
    "                else:\n",
    "                    n_temp=i.replace('deseq_','').replace('.txt','')\n",
    "                deseq.to_excel(writer, sheet_name=n_temp)\n",
    "            writer.save()\n",
    "        else:\n",
    "            raise ValueError('No DESEq result found')\n",
    "\n",
    "    def save_object(self,filename='DEseq_PIANO.pkl'):\n",
    "        with open('%s/%s' % (self.respath,filename), 'wb') as file:\n",
    "            pickle.dump(self, file)\n",
    "    \n",
    "    def load_object(self,filename='DEseq_PIANO.pkl'):\n",
    "        with open(filename, 'rb') as file:\n",
    "            res=pickle.load(file)\n",
    "        return res\n",
    "\n",
    "def calc(tpm,pval_thr=0.05,pos_only=False):\n",
    "    print('Calculating Correlation..')\n",
    "    temp=spearmanr(tpm.T)\n",
    "    corr=pd.DataFrame(temp[0],columns=list(tpm.index),index=list(tpm.index))\n",
    "    pval=pd.DataFrame(temp[1],columns=list(tpm.index),index=list(tpm.index))\n",
    "    print('Filtering the matrix Correlation..')\n",
    "    corr=corr.where(np.triu(np.ones(corr.shape)).astype(np.bool))\n",
    "    pval=pval.where(np.triu(np.ones(pval.shape)).astype(np.bool))\n",
    "    print('Making long table of Correlation..')\n",
    "    corr2=corr.unstack().reset_index(name='weight')\n",
    "    pval2=pval.unstack().reset_index(name='pval')\n",
    "    res=corr2.merge(pval2,on=['level_0','level_1'])\n",
    "    res=res[res['level_0'] != res['level_1']]\n",
    "    res=res.dropna()\n",
    "    print('Adjusting P-val')\n",
    "    res['padj']=multipletests(res['pval'],method='fdr_bh')[1]\n",
    "    res=res[res.padj < pval_thr].reset_index(drop=True)\n",
    "    if pos_only:\n",
    "        res=res[res['weight']>0]\n",
    "    res=res[['level_0','level_1','weight']]\n",
    "    print('Done!!')\n",
    "    return res\n",
    "def filter_fpkm(fpkm,conds,thr=1):\n",
    "    return fpkm[(pd.concat([fpkm.T,conds],1).groupby(conds.name).mean().T>thr).sum(1)>0]\n",
    "\n",
    "class Network_Analysis:\n",
    "    def __init__(self,filename,respath,KEGG='',GO='',topx=0.05,cluster_size=30):\n",
    "        self.network_ori=pd.read_csv(filename,sep='\\t')\n",
    "        self.res_path=respath\n",
    "        if os.path.isdir(self.res_path):\n",
    "            pass\n",
    "        else:\n",
    "            os.mkdir(self.res_path)\n",
    "        self.cluster_size=cluster_size\n",
    "        self.topx=topx\n",
    "        self.GSC_KEGG=KEGG\n",
    "        self.GSC_GO=GO\n",
    "        self.__net_analysis()\n",
    "        self.export_cluster_cytoscape()\n",
    "        print('Enriching Clusters...')\n",
    "        self.__cluster_enrichr()\n",
    "        \n",
    "    def __net_analysis(self):\n",
    "        print('Loading The Network...')\n",
    "        temp=self.network_ori[self.network_ori['weight']>0]\n",
    "        temp=temp[temp['weight']>temp['weight'].quantile(1-self.topx)]\n",
    "        g= ig.Graph.TupleList(zip(temp['level_0'],temp['level_1']))\n",
    "        print('Cluster Analysis...')\n",
    "        pd.Series(g.degree(), index = g.vs['name']).sort_values(ascending = False).to_csv('%s/degree_pos.txt' % self.res_path,sep='\\t')\n",
    "        optimiser = leidenalg.Optimiser()\n",
    "        clust_calc = leidenalg.ModularityVertexPartition(g)\n",
    "        diff =  optimiser.optimise_partition(clust_calc, n_iterations=-1)\n",
    "        clust=pd.Series(clust_calc.membership,index=g.vs['name'])\n",
    "        thr=self.cluster_size\n",
    "        temp_c=clust.value_counts()[clust.value_counts()>thr].index.tolist()\n",
    "        self.clustering=clust[clust.isin(temp_c)]\n",
    "        self.modularity=clust_calc.modularity\n",
    "        self.network=g\n",
    "        \n",
    "    def __enrichr(self,GSC,db_name):\n",
    "        final={}\n",
    "        for i in range(self.clustering.value_counts().shape[0]):\n",
    "            if db_name == 'GO':\n",
    "                gene_list=list([i.upper() for i in self.clustering[self.clustering.isin([i])].index])\n",
    "            else:\n",
    "                gene_list=list([i for i in self.clustering[self.clustering.isin([i])].index])\n",
    "            enr = gp.enrichr(gene_list=gene_list,\n",
    "                             description='test_name',\n",
    "                             gene_sets=GSC,\n",
    "                             background=20000,\n",
    "                             outdir='test/enrichr_kegg2',\n",
    "                             cutoff=0.5,\n",
    "                             verbose=False)\n",
    "            \n",
    "            try:\n",
    "                temp=enr.res2d[enr.res2d['Adjusted P-value']<0.05].sort_values('Adjusted P-value')[['Term','Adjusted P-value']]\n",
    "                temp.columns=['%s' % db_name,'P-Adj (%s)' % db_name]\n",
    "                if db_name == 'GO':\n",
    "                    temp['GOID']=[s[s.find(\"(\")+1:s.find(\")\")] for s in temp['GO']]\n",
    "                    temp=temp[['GO','GOID','P-Adj (GO)']]\n",
    "            except TypeError:\n",
    "                temp = pd.DataFrame(columns = ['Term','Adjusted P-value'])\n",
    "                if db_name == 'GO':\n",
    "                    temp = pd.DataFrame(columns = ['GO','GOID','P-Adj (GO)'])\n",
    "            final[i]=temp.reset_index(drop=True)\n",
    "        return final\n",
    "    \n",
    "    def __cluster_enrichr(self):\n",
    "        if (self.GSC_KEGG != '') | (self.GSC_GO != ''):\n",
    "            writer = pd.ExcelWriter('%s/Coexp_Cluste_Enrichr.xlsx' % self.res_path, engine='xlsxwriter')\n",
    "            if self.GSC_KEGG != '':\n",
    "                self.KEGG=self.__enrichr(self.GSC_KEGG,'KEGG')\n",
    "                pd.concat(self.KEGG.values(), axis=1, keys=self.KEGG.keys()).to_excel(writer, sheet_name='KEGG')\n",
    "\n",
    "            if self.GSC_GO != '':\n",
    "                self.GO=self.__enrichr(self.GSC_GO,'GO')\n",
    "                pd.concat(self.GO.values(), axis=1, keys=self.GO.keys()).to_excel(writer, sheet_name='GO')\n",
    "\n",
    "            if (self.GSC_KEGG != '') & (self.GSC_GO != ''):\n",
    "                self.Both={i:pd.concat([self.GO[i],self.KEGG[i]],1) for i in range(self.clustering.value_counts().shape[0])}\n",
    "                pd.concat(self.Both.values(), axis=1, keys=self.Both.keys()).to_excel(writer, sheet_name='Both')\n",
    "            writer.save()\n",
    "            shutil.rmtree('test')\n",
    "    \n",
    "    def save_network(self):\n",
    "        pickle_out = open('%s/network_object.pkl' % self.res_path,\"wb\")\n",
    "        pickle.dump(self, pickle_out)\n",
    "        pickle_out.close()\n",
    "        \n",
    "    def export_cluster_cytoscape(self):\n",
    "        temp=self.network_ori.merge(pd.DataFrame(self.clustering).reset_index(),how='left',left_on='level_0',right_on='index').merge(pd.DataFrame(self.clustering).reset_index(),how='left',left_on='level_1',right_on='index').dropna()\n",
    "        temp2=temp[['0_x','0_y']]\n",
    "        temp2['count']=1\n",
    "        temp2=temp2.groupby(['0_x','0_y']).count().reset_index().pivot_table(index='0_x',columns='0_y',values='count').fillna(0)\n",
    "        temp2=pd.DataFrame([(i,j,(temp2.loc[i,j]+temp2.loc[j,i]),(temp2.loc[i,j]+temp2.loc[j,i])/(self.clustering.value_counts()[i]+self.clustering.value_counts()[j])) for i in temp2.index for j in temp2.columns if (i < j)],columns=['Source','Target','Weight1','Weight2'])\n",
    "        temp2.astype(int).to_csv('%s/edge_pos_cluster.txt' % self.res_path,sep='\\t')\n",
    "        pd.DataFrame(self.clustering.value_counts(),columns=['Size']).to_csv('%s/node_pos_cluster.txt' % self.res_path,sep='\\t')\n",
    "        self.clustering.to_csv('%s/clust_pos_cluster.txt' % self.res_path,sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c5ff810-780e-45ee-9afa-ae32035f927d",
   "metadata": {},
   "source": [
    "## Data Loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "980ea32d-b2fe-48ee-be89-60c9a8fdb894",
   "metadata": {},
   "outputs": [],
   "source": [
    "sup1 = pd.ExcelFile('SupplementaryTables/SupplementaryTable1.xlsx')\n",
    "\n",
    "conds=sup1.parse('Metadata', index_col = 0)['conds']\n",
    "count=sup1.parse('Count', index_col = 0)[conds.index]\n",
    "fpkm=sup1.parse('FPKM', index_col = 0)[conds.index]\n",
    "experimental = sup1.parse('Experimental Variables', index_col = 0, skiprows =2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03d7f14f-b53f-4ab5-9041-f3f9f822ccfd",
   "metadata": {},
   "source": [
    "## PCA (Figure 1B)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b554f25d-7a3c-482f-a327-4f9381e28182",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = pd.concat([fpkm.T,conds],1).groupby('conds').median()\n",
    "temp = temp[temp>1].dropna(how='all',axis = 1)\n",
    "sub = fpkm.reindex(set(temp.columns)).dropna()\n",
    "sub.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c196355f-d10e-42e4-af65-fc8bdf8338fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "var = 'conds'\n",
    "sklearn_pca = sklearnPCA(n_components=2)\n",
    "Y_sklearn = sklearn_pca.fit_transform(np.log1p(sub.T))\n",
    "plt.figure(figsize=(6,5))\n",
    "coordinate = pd.DataFrame(Y_sklearn, columns = ['x', 'y'], index = conds).reset_index()\n",
    "ax = sns.scatterplot(data = coordinate, x = 'x', y = 'y', s = 300, hue = 'conds')\n",
    "ax.set_xlabel('PC1 (%.2f%%)' % (sklearn_pca.explained_variance_ratio_[0]*100), fontsize = 20, fontweight = 'bold')\n",
    "ax.set_ylabel('PC2 (%.2f%%)' % (sklearn_pca.explained_variance_ratio_[1]*100), fontsize = 20, fontweight = 'bold')\n",
    "ax.set_xticklabels([])\n",
    "ax.set_yticklabels([])\n",
    "ax.get_legend().set_title('Age Group')\n",
    "plt.setp(ax.get_legend().get_title(), fontsize='15', fontweight = 'bold')\n",
    "plt.setp(ax.get_legend().get_texts(), fontsize='15')\n",
    "plt.tight_layout()\n",
    "plt.savefig('Figures/1B.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "affd9ec7-8b3d-4692-835a-82c13b8a73be",
   "metadata": {},
   "source": [
    "## Differential Expression Analysis\n",
    "The results of this analysis are under the folder in res_path. THis is associated with DEG, GO, and KEGG in Supplementary Table 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e388684f-a24a-44e5-bce4-349ca8253ba8",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_path='DiffExp_FunctionalAnalysis/'\n",
    "k=DESeq_Piano(res_path)\n",
    "k.DEseq(count,conds)\n",
    "k.loadGSC(KEGG='lib/KEGG_rat_Novogene.gmt', GO='lib/GO_Biological_Process_2021.gmt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87dd662d-b607-4c62-808a-51ce2b10ed9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "cond1='Aging'\n",
    "cond2='Young'\n",
    "k.DEseq_Compare(cond1,cond2,save=True)\n",
    "k.PIANO(cond1,cond2,heatmap=False,save=True)\n",
    "\n",
    "k.summarizeDEseq()\n",
    "k.summarizePiano()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25281bc1-9188-41a2-b6e5-285917bd58ef",
   "metadata": {},
   "source": [
    "## Figure 1C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "892527d1-e726-471d-b0f2-abc887d6b0dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = pd.ExcelFile('DiffExp_FunctionalAnalysis/DifferentialExpression_ALL.xlsx')\n",
    "res = []\n",
    "for i in temp.sheet_names[0:]:\n",
    "    ha = temp.parse(i, index_col = 0)\n",
    "    ha = ha[ha['P-ADJ'] < 0.05]['DIRECTION']\n",
    "    res.append(ha.rename(i.replace('_', ' vs ')))\n",
    "res = pd.concat(res, 1)\n",
    "temp2 = res.unstack().reset_index().dropna().groupby(['level_0',0]).count().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10526b52-ecc5-49a5-96b7-204e44c4103a",
   "metadata": {},
   "outputs": [],
   "source": [
    "res.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ded0cd1-cdef-481d-8d15-6f53c95badcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize = (2,4))\n",
    "g = sns.barplot(data = temp2, x = 0, y = 'rownames', order = ['UP', 'DOWN'], palette = [\"#a50000\", \"#0000a5\"])\n",
    "# g.set_yticklabels(g.get_yticklabels(), fontsize = 10)\n",
    "g.set_xticklabels(g.get_xticklabels(), fontsize = 15, fontweight = 'bold')\n",
    "g.set_xlabel('')\n",
    "g.set_ylabel('')\n",
    "plt.tight_layout()\n",
    "plt.savefig('Figures/1C.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72bd1fa9-77b5-4e06-89b2-d66bd5004e13",
   "metadata": {},
   "source": [
    "## Figure 1D & 2F"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a0d7fb5-1254-4e37-8647-955cef0d2668",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = pd.ExcelFile('DiffExp_FunctionalAnalysis/GO_ALL.xlsx')\n",
    "res = []\n",
    "for i in temp.sheet_names:\n",
    "    ha = temp.parse(i, index_col = 0)\n",
    "    ha_up = ha[ha['Stats'] > 0]\n",
    "    ha_up = ha_up.replace(0, ha_up[ha_up['P-Adj'] > 0]['P-Adj'].min()*0.01)\n",
    "    ha_dn = ha[ha['Stats'] < 0]\n",
    "    ha_dn = ha_dn.replace(0, ha_dn[ha_dn['P-Adj'] > 0]['P-Adj'].min()*0.01)\n",
    "    ha = pd.concat([ha_up, ha_dn])\n",
    "    ha[i.replace('_', ' vs ')] = (-np.log10(ha['P-Adj'])*np.sign(ha['Stats']))\n",
    "    res.append(ha[i.replace('_', ' vs ')])\n",
    "res = pd.concat(res, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b35b5fa-fe25-4557-b821-a896396f5b5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_sel = res[res.index.str.contains('mitoch')].reset_index()\n",
    "res_sel['Name'] = [i.split(' (GO')[0].replace('mitochondrial', 'MT').replace('electron transport', 'ET').replace('respiratory chain', 'RC') for i in res_sel['Name']]\n",
    "plt.figure(figsize = (2, 4))\n",
    "ax = sns.barplot(data = res_sel, x = 'Aging vs Young', y = 'Name', color = \"#0000a5\")\n",
    "ax.set_ylabel('')\n",
    "ax.set_xlabel('Gene-Level Statistics', fontsize = 20, fontweight = 'bold')\n",
    "ax.vlines(0, ymin = ax.get_ylim()[1], ymax = ax.get_ylim()[0], color = 'k')\n",
    "ax.set_yticklabels(ax.get_yticklabels(), fontsize = 15, )\n",
    "sns.despine(ax=ax, top=True, right=True, left=True, bottom=False)\n",
    "plt.savefig('Figures/1D.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbfaccda-9594-4c09-9c29-05d5e984e867",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_sel = res[res.index.str.contains('fatty') | res.index.str.contains('lipopro') | res.index.str.contains('chylomic') | res.index.str.contains('lipid ')].reset_index()\n",
    "res_sel = res_sel[~res_sel['Name'].str.contains('negative') & ~res_sel['Name'].str.contains('positive')]\n",
    "res_sel['Name'] = [i.split(' (GO')[0].replace('fatty acid', 'FA').replace('very-low-density lipoprotein', 'vLDL').replace('low-density lipoprotein', 'LDL').replace('high-density lipoprotein', 'HDL').replace('triglyceride', 'TG') for i in res_sel['Name']]\n",
    "res_sel = res_sel.reindex([2, 3, 6, 9, 11, 13, 14, 16, 19]).sort_values('Aging vs Young') # Selecting to include just the upper level of GO\n",
    "\n",
    "plt.figure(figsize = (4, 3))\n",
    "ax = sns.barplot(data = res_sel, x = 'Aging vs Young', y = 'Name', color = \"#0000a5\")\n",
    "ax.set_ylabel('')\n",
    "ax.set_xlabel('Gene-Level Statistics', fontsize = 20, fontweight = 'bold')\n",
    "ax.vlines(0, ymin = ax.get_ylim()[1], ymax = ax.get_ylim()[0], color = 'k')\n",
    "ax.set_yticklabels(ax.get_yticklabels(), fontsize = 15, )\n",
    "sns.despine(ax=ax, top=True, right=True, left=True, bottom=False)\n",
    "plt.savefig('Figures/2F.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3f8f8dc-6971-4594-bfab-358c867997a7",
   "metadata": {},
   "source": [
    "## Figure 1E and 2A-E\n",
    "Generating the figures, then post-processing was done using graphic software"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92352199-1fb9-4faa-9165-ab557c0e9786",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Excluding the unrelated diseases and pathways\n",
    "exclude_KEGG = ['Hepatitis B', 'Human T-cell leukemia virus 1 infection', 'Regulation of actin cytoskeleton', 'Measles',\n",
    "           'MicroRNAs in cancer', 'Pathways in cancer', 'Influenza A', 'Oocyte meiosis', 'Progesterone-mediated oocyte maturation',\n",
    "           'Ribosome biogenesis in eukaryotes', 'Amoebiasis', 'Small cell lung cancer', 'Epstein-Barr virus infection', 'Human papillomavirus infection',\n",
    "           'Pertussis', 'Staphylococcus aureus infection', 'Proteoglycans in cancer', 'Leishmaniasis', 'Legionellosis', 'Viral carcinogenesis', \n",
    "           'Chagas disease (American trypanosomiasis)', 'Toxoplasmosis', 'Malaria', 'Metabolism of xenobiotics by cytochrome P450', 'Chronic myeloid leukemia',\n",
    "           'Prion diseases', 'Inflammatory bowel disease (IBD)', 'Kaposi sarcoma-associated herpesvirus infection', 'Parkinson disease', 'Hepatitis C', 'Systemic lupus erythematosus',\n",
    "           'Tuberculosis', 'Hepatitis B', 'Measles', 'Dilated cardiomyopathy (DCM)', 'Hematopoietic cell lineage', 'Adrenergic signaling in cardiomyocytes',\n",
    "           'Epstein-Barr virus infection', 'Drug metabolism', 'Salivary secretion', 'Thermogenesis', 'Glioma', 'Human cytomegalovirus infection', 'Melanoma', 'Chemical carcinogenesis',\n",
    "            'African trypanosomiasis', 'Collecting duct acid secretion', 'African trypanosomiasis', 'Vascular smooth muscle contraction', 'Fluid shear stress and atherosclerosis',\n",
    "            'Rheumatoid arthritis', 'Metabolic pathways', 'Spinocerebellar ataxia', 'Drug metabolism - other enzymes', 'Asthma'\n",
    "          ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c6b54a6-ecd9-439a-bd39-68043b6a0ca0",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = open('lib/ko00001.keg', 'r').readlines()\n",
    "x = []\n",
    "for i in temp:\n",
    "    if i.startswith('A'):\n",
    "        A = ' '.join(i.split(' ')[1:]).rstrip('\\n')\n",
    "    if i.startswith('B  '):\n",
    "        B = ' '.join(i.replace('B  ', '').split(' ')[1:]).rstrip('\\n')\n",
    "    if i.startswith('C    '):\n",
    "        C = ' '.join(i.replace('C    ', '').split(' ')[1:]).split(' [PATH')[0].rstrip('\\n')\n",
    "        x.append([A, B, C])\n",
    "x = pd.DataFrame(x, columns = ['A', 'B', 'C'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b02ffbb-cc0c-4b19-9bf9-9bf19c9286d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = pd.ExcelFile('DiffExp_FunctionalAnalysis/KEGG_ALL.xlsx')\n",
    "res = []\n",
    "for i in temp.sheet_names:\n",
    "    ha = temp.parse(i, index_col = 0)\n",
    "    ha_up = ha[ha['Stats'] > 0]\n",
    "    ha_up = ha_up.replace(0, ha_up[ha_up['P-Adj'] > 0]['P-Adj'].min()*0.01)\n",
    "    ha_dn = ha[ha['Stats'] < 0]\n",
    "    ha_dn = ha_dn.replace(0, ha_dn[ha_dn['P-Adj'] > 0]['P-Adj'].min()*0.01)\n",
    "    ha = pd.concat([ha_up, ha_dn])\n",
    "    ha[i.replace('_', ' vs ')] = (-np.log10(ha['P-Adj'])*np.sign(ha['Stats']))\n",
    "    res.append(ha[i.replace('_', ' vs ')])\n",
    "res = pd.concat(res, 1)\n",
    "# Excluding the unrelated diseases and pathways\n",
    "res = res[~(res.index.str.contains('isease') | res.index.str.contains('nfection') | res.index.str.contains('ancer')| res.index.str.contains('epatitis')\n",
    "           | res.index.str.contains('ardio') | res.index.str.contains('Malaria')  | res.index.str.contains('Gastric')  | res.index.str.contains('Influenza')\n",
    "            | res.index.str.contains('Leishmaniasis') | res.index.str.contains('cleros') \n",
    "            | res.index.isin(exclude_KEGG)\n",
    "           )]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a62ba774-cb6e-4d04-8e04-6f2b4d059e16",
   "metadata": {},
   "outputs": [],
   "source": [
    "var = 'Aging vs Young'\n",
    "temp = res[var].rename('Gene Level Statistics').dropna().sort_values().reset_index()\n",
    "temp['Dir'] = ['#0000a5' if i < 0 else '#a50000' for i in temp['Gene Level Statistics']]\n",
    "temp = temp.merge(x, left_on = 'Name', right_on = 'C').sort_values(['A', 'B', 'Gene Level Statistics'])\n",
    "N = len(temp['B'].unique())\n",
    "color_mapping = dict(zip(sorted(temp['B'].unique()),plt.get_cmap('gist_rainbow')(np.linspace(0,1,N))))\n",
    "temp['col'] = [color_mapping[i] for i in temp['B']]\n",
    "\n",
    "heights = [temp[temp['B'] == i].shape[0] for i in temp['B'].unique()]\n",
    "fig, ax = plt.subplots(nrows=N, ncols=1, figsize = (4, 40), sharex = True,\n",
    "                       gridspec_kw={\n",
    "                           'height_ratios': heights,\n",
    "                       'wspace': 0.4,\n",
    "                       'hspace': 0.4})\n",
    "\n",
    "\n",
    "for row,i in enumerate(temp['B'].unique()):\n",
    "    print(i)\n",
    "    temp2 = temp[temp['B'] == i]\n",
    "    temp2['Name'] = temp2['Name'].str.replace(' pathway', '')\n",
    "    sns.barplot(data = temp2, y = 'Name', x = 'Gene Level Statistics', palette = temp2['Dir'], ax = ax[row])\n",
    "    ax[row].set_title(i, fontsize = 15, fontweight = 'bold')\n",
    "    ax[row].set_ylabel('')\n",
    "    ax[row].vlines(0, ymin = ax[row].get_ylim()[1], ymax = ax[row].get_ylim()[0], color = 'k')\n",
    "#     ax[row].set_xticklabels(ax[row].get_xticklabels(), fontsize = 15, fontweight = 'bold')\n",
    "    ax[row].set_yticklabels(ax[row].get_yticklabels(), fontsize = 15, )\n",
    "    if row != (len(temp['B'].unique()) - 1):\n",
    "        ax[row].set_xlabel('')\n",
    "        sns.despine(ax=ax[row], top=True, right=True, left=True, bottom=True)\n",
    "    else:\n",
    "        sns.despine(ax=ax[row], top=True, right=True, left=True, bottom=False)\n",
    "plt.tight_layout()\n",
    "plt.savefig('Figures/KEGG_1E_2A-E.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "158ac64b-e16f-42ce-816e-2e8491a293b2",
   "metadata": {},
   "source": [
    "## Figure 1F-H, 1J, and 4D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca2febcd-a747-4140-a3d3-d833fa2faf2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "var_name = {\n",
    "    'CI': '1F',\n",
    "    'CII': '1G',\n",
    "    'CIV': '1H',\n",
    "    'NAD': '1J',\n",
    "    'TotalCholesterol': '4D',\n",
    "}\n",
    "\n",
    "for var in list(var_name.keys())[0:-1]:\n",
    "    plt.figure(figsize = (2.5,4))\n",
    "    fc = pd.concat([experimental[var]/experimental[experimental['Age'] == 'Y'][var].mean(), experimental['Age']],1)\n",
    "    order = ['Y', 'A']\n",
    "    box_pairs = [\n",
    "        ('Y', 'A'),\n",
    "    ]\n",
    "    g = sns.barplot(data = fc, x = 'Age', y = var, order=order, palette = ['white', '#a50000'], capsize=.2,  edgecolor='black')#, hue = 'Gender', hue_order = hue_order)#, ax = axs[num])\n",
    "    sns.stripplot(data = fc[fc['Age'] == 'Y'], x = 'Age', y = var, order=order, marker = 'o', color = 'black', alpha = 0.75)#, dodge = False, hue = 'Gender', hue_order = hue_order)#, ax = axs[num])\n",
    "    sns.stripplot(data = fc[fc['Age'] == 'A'], x = 'Age', y = var, order=order, marker = 's', color = 'black', alpha = 0.75)\n",
    "    add_stat_annotation(data = fc, x = 'Age', y = var, order=order, ax = g,#, hue_order = hue_order\n",
    "                        box_pairs=box_pairs,\n",
    "                        perform_stat_test=True, #pvalues=pval_sel,\n",
    "                        test='t-test_ind', text_format='star', loc='inside', verbose=0, fontsize = 15)\n",
    "    g.set_ylabel('%s FC' % var)\n",
    "    plt.tight_layout()\n",
    "    plt.savefig('Figures/%s.pdf' % var_name[var])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fd6cfe8-f282-4682-98a1-3198274f2ea4",
   "metadata": {},
   "outputs": [],
   "source": [
    "var = 'TotalCholesterol'\n",
    "plt.figure(figsize = (2.5,4))\n",
    "order = ['Y', 'A']\n",
    "box_pairs = [\n",
    "    ('Y', 'A'),\n",
    "]\n",
    "g = sns.barplot(data = experimental, x = 'Age', y = var, order=order, palette = ['white', '#a50000'], capsize=.2,  edgecolor='black')#, hue = 'Gender', hue_order = hue_order)#, ax = axs[num])\n",
    "sns.stripplot(data = experimental[experimental['Age'] == 'Y'], x = 'Age', y = var, order=order, marker = 'o', color = 'black', alpha = 0.75)#, dodge = False, hue = 'Gender', hue_order = hue_order)#, ax = axs[num])\n",
    "sns.stripplot(data = experimental[experimental['Age'] == 'A'], x = 'Age', y = var, order=order, marker = 's', color = 'black', alpha = 0.75)\n",
    "add_stat_annotation(data = experimental, x = 'Age', y = var, order=order, ax = g,#, hue_order = hue_order\n",
    "                    box_pairs=box_pairs,\n",
    "                    perform_stat_test=True, #pvalues=pval_sel,\n",
    "                    test='t-test_ind', text_format='star', loc='inside', verbose=0, fontsize = 15)\n",
    "plt.tight_layout()\n",
    "plt.savefig('Figures/%s.pdf' % var_name[var])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eded6115-aa5d-4414-807f-f24be9dfc237",
   "metadata": {},
   "source": [
    "## Figure 3A-D and Supplementary Data 2\n",
    "The raw data was downloaded from SRA database. Use the accession number to download.\n",
    "\n",
    "Differential Expression analysis was performed in the same way as above. \n",
    "\n",
    "Venn diagram was created using graphic software\n",
    "\n",
    "Functional analysis was done using Gseapy Enrichr API with the the KEGG library (lib/KEGG_2021_Human.gmt)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ff26a28-d54e-4fa5-ac39-18f980efa8e8",
   "metadata": {},
   "source": [
    "## Network Generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea69ca43-5f44-41af-9e0a-26c6400deb44",
   "metadata": {},
   "outputs": [],
   "source": [
    "coexp = calc(filter_fpkm(fpkm, conds, thr = 1))\n",
    "coexp.to_csv('Network/coexp.txt', sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f06df6ef-40b4-41ee-9960-2b60ecb6a079",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This was repeated 100 times to get the most robust clustering\n",
    "topx = 0.1\n",
    "k=Network_Analysis(filename='Network/coexp.txt',respath='Network/Analysis/',KEGG='lib/KEGG_rat_Novogene.gmt',GO='lib/GO_Biological_Process_2021.gmt',topx=topx)\n",
    "k.save_network()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "017547a9-151e-4d92-a5c9-de6b2762ff51",
   "metadata": {},
   "source": [
    "## Network Analysis (Figure 4A-C and Supplemtentary Data 3)\n",
    "\n",
    "This figures were generated with Cytoscape. Both Cytoscape Files are available under \"Network\" folder\n",
    "\n",
    "Figure 4A: Use the '{edges/node}_pos_cluster.txt'. The transparency of the edges were proportional to the weight and size of the nodes is proportional to the average clustering coefficient (calculation below, related file \"avgClustCoeff.txt\")\n",
    "\n",
    "Figure 4B: Use the subnodes and edges from the calculation below. Color was based on the differential expression results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b744a31e-71a5-41ff-a4dc-01744d60015c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Loading deseq info\n",
    "temp = pd.ExcelFile('DiffExp_FunctionalAnalysis/DifferentialExpression_ALL.xlsx')\n",
    "res = []\n",
    "for i in temp.sheet_names[0:]:\n",
    "    ha = temp.parse(i, index_col = 0)\n",
    "    ha = ha[ha['P-ADJ'] < 0.05]['DIRECTION']\n",
    "    res.append(ha.rename(i.replace('_', ' vs ')))\n",
    "res = pd.concat(res, 1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06f540e2-4d4d-4006-9af8-d4c3d716277b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Average clustering coefficient\n",
    "k = pickle.load( open( 'Network/Analysis/network_object.pkl', \"rb\" ) )\n",
    "x = []\n",
    "\n",
    "for i in sorted(k.clustering.unique()):\n",
    "    x.append([i, k.network.subgraph(k.clustering[k.clustering == i].index).transitivity_undirected()])\n",
    "avgClustCoeff = pd.DataFrame(x, columns = ['Cluster','AvgClustCoeff']).sort_values('AvgClustCoeff', ascending = False)\n",
    "avgClustCoeff.to_csv('Network/Analysis/avgClustCoeff.txt', sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d79b605-9805-4053-8979-0d68d917bbe8",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = pd.concat([k.clustering, res], 1)\n",
    "cluster_deseq = temp[temp[0].notna()]\n",
    "cluster_deseq['count'] = 1\n",
    "cluster_deseq.fillna('Unchanged').groupby([0, 'Aging vs Young']).count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d522ec4-f079-4a70-820c-75298ace1cc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Subnetwork Cluster 14 (Figure 5B)\n",
    "degree = pd.Series(k.network.degree(), index = k.network.vs['name'])\n",
    "\n",
    "for cl in k.clustering.unique():\n",
    "    nodes = k.clustering[k.clustering == cl]\n",
    "    nodes = pd.concat([nodes, res.reindex(nodes.index), degree.reindex(nodes.index)],1)\n",
    "    temp = k.network.subgraph(nodes.index)\n",
    "    edges = []\n",
    "    for edge in temp.es:\n",
    "        source_vertex_id = edge.source\n",
    "        target_vertex_id = edge.target\n",
    "        source_vertex = temp.vs[source_vertex_id]['name']\n",
    "        target_vertex = temp.vs[target_vertex_id]['name']\n",
    "        edges.append([source_vertex, target_vertex])\n",
    "    edges = pd.DataFrame(edges, columns = ['source','target'])\n",
    "\n",
    "    edges.to_csv('Network/Analysis/subnetwork/cl_%d_edges.txt' % cl, sep = '\\t')\n",
    "    nodes.to_csv('Network/Analysis/subnetwork/cl_%d_nodes.txt' % cl, sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04c84801-d42d-4fc4-9802-88c65ac523b7",
   "metadata": {},
   "source": [
    "## Figure 5 and Supplementary Data 4\n",
    "\n",
    "Figure 5B-D and Supplementary Data 4: Originated from public GTEx data. Selected samples can be found in Supplementary Data 4 and the flow to select them is presented in Supplementary Figure 4. Differential Expression and Functional Analysis were same as above. \n",
    "\n",
    "Figure 5E-F: Single cell coordinates and expression values were extracted from Seurat object shared by the original authors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4315f5fe-6efb-4049-98db-486631987310",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Figure 5A\n",
    "temp = pd.concat([fpkm.loc['Pcsk9'], conds],1)\n",
    "plt.figure(figsize = (2, 4))\n",
    "g = sns.barplot(data = temp, x = 'conds', y = 'Pcsk9', order=['Young', 'Aging'],palette = ['white', '#a50000'], capsize=.2,  edgecolor='black')#, hue = 'Gender', hue_order = hue_order)#, ax = axs[num])\n",
    "plt.savefig('Figures/5A.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16de11fc-c3ab-4214-81fa-565134e2094a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Figure 6C-D\n",
    "\n",
    "exclude_KEGG = ['Hepatitis B', 'Human T-cell leukemia virus 1 infection', 'Regulation of actin cytoskeleton', 'Measles',\n",
    "           'MicroRNAs in cancer', 'Pathways in cancer', 'Influenza A', 'Oocyte meiosis', 'Progesterone-mediated oocyte maturation',\n",
    "           'Ribosome biogenesis in eukaryotes', 'Amoebiasis', 'Small cell lung cancer', 'Epstein-Barr virus infection', 'Human papillomavirus infection',\n",
    "           'Pertussis', 'Staphylococcus aureus infection', 'Proteoglycans in cancer', 'Leishmaniasis', 'Legionellosis', 'Viral carcinogenesis', \n",
    "           'Chagas disease (American trypanosomiasis)', 'Toxoplasmosis', 'Malaria', 'Metabolism of xenobiotics by cytochrome P450', 'Chronic myeloid leukemia',\n",
    "           'Prion diseases', 'Inflammatory bowel disease (IBD)', 'Kaposi sarcoma-associated herpesvirus infection', 'Parkinson disease', 'Hepatitis C', 'Systemic lupus erythematosus',\n",
    "           'Tuberculosis', 'Hepatitis B', 'Measles', 'Dilated cardiomyopathy (DCM)', 'Hematopoietic cell lineage', 'Adrenergic signaling in cardiomyocytes',\n",
    "           'Epstein-Barr virus infection', 'Drug metabolism', 'Salivary secretion', 'Thermogenesis', 'Glioma', 'Human cytomegalovirus infection', 'Melanoma', 'Chemical carcinogenesis',\n",
    "            'African trypanosomiasis', 'Collecting duct acid secretion', 'African trypanosomiasis', 'Vascular smooth muscle contraction', 'Fluid shear stress and atherosclerosis',\n",
    "            'Rheumatoid arthritis', 'Metabolic pathways', 'Spinocerebellar ataxia', 'Drug metabolism - other enzymes', 'Asthma', 'Allograft rejection', 'Type I diabetes mellitus'\n",
    "          ]\n",
    "\n",
    "KEGG_ourdata = sup1.parse('KEGG Liver', index_col = 0)['Stats'].rename('Aging vs Young')\n",
    "GO_ourdata = sup1.parse('GO BP Liver', index_col = 0)['Stats'].rename('Aging vs Young')\n",
    "\n",
    "\n",
    "sup4 = pd.ExcelFile('SupplementaryTables/SupplementaryTable4.xlsx')\n",
    "KEGG_Gtex = sup4.parse('KEGG GTEx', index_col = 0)['Stats'].rename('High vs Low\\nPCSK9 (GTEx)')\n",
    "GO_Gtex = sup4.parse('GO BP GTEx', index_col = 0)['Stats'].rename('High vs Low\\nPCSK9 (GTEx)')\n",
    "\n",
    "KEGG = pd.concat([KEGG_ourdata, KEGG_Gtex], 1).dropna()\n",
    "KEGG = KEGG[~(KEGG.index.str.contains('isease') | KEGG.index.str.contains('nfection') | KEGG.index.str.contains('ancer')| KEGG.index.str.contains('epatitis')\n",
    "           | KEGG.index.str.contains('ardio') | KEGG.index.str.contains('Malaria')  | KEGG.index.str.contains('Gastric')  | KEGG.index.str.contains('Influenza')\n",
    "            | KEGG.index.str.contains('Leishmaniasis') | KEGG.index.str.contains('cleros') \n",
    "            | KEGG.index.isin(exclude_KEGG)\n",
    "           )]\n",
    "GO = pd.concat([GO_ourdata, GO_Gtex], 1).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3ae6297-b7ec-47f7-ab07-3b8ce8df2aae",
   "metadata": {},
   "outputs": [],
   "source": [
    "cmap=matplotlib.colors.LinearSegmentedColormap.from_list(\"\", [\"#0000a5\",'#0000d8',\"#FFFAF0\",'#d80000',\"#a50000\"])\n",
    "g = sns.clustermap(KEGG, yticklabels = 1, \n",
    "               center = 0, cmap = cmap, \n",
    "               col_cluster = False, row_cluster = True,\n",
    "               figsize = (5,7),\n",
    "               cbar_kws={\"orientation\": \"vertical\",\"shrink\": 0.8, 'label': 'Gene-Level Statistics'}\n",
    "              )\n",
    "g.ax_heatmap.set_xticklabels(g.ax_heatmap.get_xticklabels(),rotation=30, ha = 'right', fontsize = 15, fontweight = 'bold')\n",
    "g.ax_heatmap.set_yticklabels(g.ax_heatmap.get_yticklabels(), fontsize = 15, )\n",
    "g.ax_heatmap.set_ylabel('')\n",
    "g.ax_heatmap.tick_params(right=False, bottom=False)\n",
    "g.savefig('Figures/5C.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "273eb649-30fb-4019-87ba-7a773ccf5ea4",
   "metadata": {},
   "outputs": [],
   "source": [
    "cmap=matplotlib.colors.LinearSegmentedColormap.from_list(\"\", [\"#0000a5\",'#0000d8',\"#FFFAF0\",'#d80000',\"#a50000\"])\n",
    "temp = GO[(GO.index.str.contains('chol')) | (GO.index.str.contains('fatty')) | (GO.index.str.contains('lipo')) | (GO.index.str.contains('lipid')) | (GO.index.str.contains('mitochondria'))]\n",
    "g = sns.clustermap(temp, yticklabels = 1, \n",
    "               center = 0, cmap = cmap, \n",
    "               col_cluster = False, row_cluster = True,\n",
    "               figsize = (5,3),\n",
    "               cbar_kws={\"orientation\": \"vertical\",\"shrink\": 0.8, 'label': 'Gene-Level Statistics'}\n",
    "              )\n",
    "g.ax_heatmap.set_xticklabels(g.ax_heatmap.get_xticklabels(),rotation=30, ha = 'right', fontsize = 15, fontweight = 'bold')\n",
    "g.ax_heatmap.set_yticklabels(g.ax_heatmap.get_yticklabels(), fontsize = 15, )\n",
    "g.ax_heatmap.set_ylabel('')\n",
    "g.ax_heatmap.tick_params(right=False, bottom=False)\n",
    "g.savefig('Figures/5D.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef2e655a-b823-4479-b07f-ff6ac8def126",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Table 1 and Supplementary Data 4: Selection of Genes\n",
    "sup2 = pd.ExcelFile('SupplementaryTables/SupplementaryTable2.xlsx')\n",
    "clusters =pickle.load(open( 'Network/Analysis/network_object.pkl', \"rb\" )).clustering\n",
    "clusters.index = clusters.index.str.upper().rename('Network Clusters')\n",
    "\n",
    "deg_our = sup1.parse('DEGs Liver', index_col = 0)\n",
    "deg_our = deg_our[deg_our['P-ADJ'] < 0.05]['DIRECTION'].rename('OwnData')\n",
    "deg_our.index = deg_our.index.str.upper()\n",
    "\n",
    "deg_24 = sup2.parse('DEGs PRJNA516151 (24Mo vs 6Mo)', index_col = 0)\n",
    "deg_24 = deg_24[deg_24['P-ADJ'] < 0.05]['DIRECTION'].rename('24')\n",
    "deg_24.index = deg_24.index.str.upper()\n",
    "\n",
    "deg_human = sup2.parse('DEGs GSE183915', index_col = 0)\n",
    "deg_human = deg_human[deg_human['P-VALUE'] < 0.05]['DIRECTION'].rename('Human')\n",
    "\n",
    "temp = pd.concat([deg_our, deg_24, deg_human, clusters],1).dropna()#.shape\n",
    "selected = temp[(temp['OwnData'] == temp['24']) & (temp['OwnData'] == temp['Human'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eba6c210-887e-41a2-9d0e-0add7ff5d875",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "LiverAging_Analysis",
   "language": "python",
   "name": "liveraging_analysis"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
